common/src/main/java/org/conscrypt/ProvidedSessionDecorator.java

```diff
@@ -18,17 +18,23 @@
 
 import java.security.Principal;
 import java.security.cert.Certificate;
+import java.util.HashMap;
 import java.util.List;
 import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.SSLSessionBindingEvent;
+import javax.net.ssl.SSLSessionBindingListener;
 import javax.net.ssl.SSLSessionContext;
 import javax.security.cert.X509Certificate;
 
 /**
  * A {@link SessionDecorator} that externalizes the provider of the delegate session. This allows
- * the underlying session to be changed externally.
+ * the underlying session to be changed externally.  This class implements the application-level
```

Comment:
I feel like the class level comment could use some love.  Maybe something like:

```java
/**
 * An externalized view of the underlying {@link SSLSession} used within a
 * socket/engine. This class provides the caller with a consistent session
 * handle which will continue to be useable regardless of internal changes
 * to the connection.  It does this by delegating calls to the <b>current</b>
 * internal session, which is provided by the session {@code Provider}
 * (i.e. the socket or engine that owns the session).
 * 
 * <p>For example, lets assume that the caller has previously obtained a
 * session object via a call to {@code getSession()}. Some time later, the caller
 * then tries to use that session but by that time the connection has been
 * closed. If the caller was using the active session directly, any calls to access
 * the the native code may result in a {@link NullPointerException} because the
 * underlying SSL object is {@code null}. However, the provider
 * (i.e. socket/engine) of the internal session in this case would provide a
 * `closed` session, which would not attempt to invoke JNI. Because of this,
 * the caller is free to use the external session at any time. 
 * 
 * <p>This class implements the {@link SSLSession} value API on behalf of the
 * internal session. This provides the caller with a consistent value map,
 * regardless of which internal session is currently being used by the
 * socket/engine.
 */
```

Reply:
Good call, expanded it out.